src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/HTTPOpenIdAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+                );
+            }
+
+            try {
+
+                UserInfoResponse userInfoResponse = UserInfoResponse.parse(httpResponse);
+
+                if (!userInfoResponse.indicatesSuccess()) {
+                    throw new AuthenticatorUnavailableException(
+                        "Error while getting " + this.userInfoEndpoint + ": " + userInfoResponse.toErrorResponse()
+                    );
+                }
+
+                String contentType = String.valueOf(httpResponse.getHeaderValues("content-type"));
+
+                JWTClaimsSet claims;
+                boolean isSigned = contentType.contains(ContentType.APPLICATION_JWT.toString());
+                if (isSigned) { // We don't need the userinfo_encrypted_response_alg since the
+                    // selfRefreshingKeyProvider has access to the keys
+                    claims = openIdJwtAuthenticator.getJwtClaimsSetFromInfoContent(
+                        userInfoResponse.toSuccessResponse().getUserInfoJWT().getParsedString()
+                    );
+                } else {
+                    claims = JWTClaimsSet.parse(userInfoResponse.toSuccessResponse().getUserInfo().toString());
+                }
+
+                String id = openIdJwtAuthenticator.getJwtClaimsSet(request).getSubject();
+                String missing = validateResponseClaims(claims, id, isSigned);
+                if (!missing.isBlank()) {
+                    throw new AuthenticatorUnavailableException(
+                        "Error while getting " + this.userInfoEndpoint + ": Missing or invalid required claims in response: " + missing
+                    );
+                }
+
+                final String subject = openIdJwtAuthenticator.extractSubject(claims);
+                if (subject == null) {
+                    log.error("No subject found in JWT token");
+                    return null;
+                }
+
+                final String[] roles = openIdJwtAuthenticator.extractRoles(claims);
+
+                AuthCredentials ac = new AuthCredentials(subject, roles);
+
+                for (Map.Entry<String, Object> claim : claims.getClaims().entrySet()) {
+                    ac.addAttribute("attr.jwt." + claim.getKey(), String.valueOf(claim.getValue()));
+                }
+
+                return ac.markComplete();
+            } catch (ParseException e) {
+                throw new RuntimeException(e);
+            }
+        } catch (IOException | URISyntaxException | com.nimbusds.oauth2.sdk.ParseException e) {
+            throw new AuthenticatorUnavailableException("Error while getting " + this.userInfoEndpoint + ": " + e, e);
+        }
+    }
+
+    private String validateResponseClaims(JWTClaimsSet claims, String id, boolean isSigned) {
+
+        String missing = "";
+
+        if (claims.getClaim(SUB_CLAIM) == null || claims.getClaim(SUB_CLAIM).toString().isBlank() || !claims.getClaim("sub").equals(id)) {
```

Comment:
This should also support the `subject_key` and `roles_key` settings: https://opensearch.org/docs/latest/security/authentication-backends/openid-connect/#configure-openid-connect-integration